### PR TITLE
fix(validation): gate generated-artifact staleness in pre_pr

### LINKED ISSUE #5079: pre_pr.py does not run build_all.py --check, so generated-file staleness reaches CI unseen
Measured during PR #5059 (2026-08-15). Cost: one full fix round, and it nearly shipped a no-op feature.

## The gap

CI runs the staleness check at `.github/workflows/validate-generated-agents.yml:158`:

    uv run python build/scripts/build_all.py --check

`scripts/validation/pre_pr.py` never runs it. The only build_all reference in the local sequence is `scripts/validation/pre_pr_sequence.py:171`:

    return validate_no_orphaned_build_deferrals(
        repo_root / "build" / "scripts" / "build_all.py", ...

That validates deferral comments *inside build_all.py's source*. It does not check whether generated output is stale relative to its sources. Different question entirely.

## What it cost

On PR #5059 the round-cap wiring was hand-edited into `src/copilot-cli/skills/pr-autofix/SKILL.md`, which is generated from `.claude/commands/pr-autofix.md`. Local runs were all green:

- `uv run pytest tests/skills/pr-autofix/` - 26/26 passed
- `uv run python build/generate_agents.py --validate` - PASSED (different pair; does not cover skills)
- `pre_pr.py` - reported only mypy and taste-ratchet issues

The work was reported as ready for PR on that evidence. CI's Build-all staleness check then failed, and the diff showed the generator stripping all 43 lines of the wiring. The feature would have merged as a silent no-op: script present, tests green, gate never invoked. That is the failure class `.claude/rules/release-it.md` describes under Silent API Migration Failures.

## Fix

Add `build_all.py --check` to the pre_pr sequence so the local gate covers the same claim CI does. Note the ordering constraint in `.claude/rules/generated-artifacts.md`: `sync_plugin_lib.py` must run before `build_all.py`, or the Copilot lib mirror is checked against stale input.

Add a test asserting the sequence includes the staleness check, so the gap cannot silently reopen.

### PR DESCRIPTION
**Pull request by rjmurillo**
# Pull Request

### What

Adds a `Generated Artifact Staleness` gate to the pre-PR validation sequence. The gate runs `scripts/sync_plugin_lib.py --check` then `build/scripts/build_all.py --check`, in the order `.claude/rules/generated-artifacts.md` requires, stopping at the first failure. Both run read-only, inside one aggregate 420-second gate budget with cleanup-preserving termination, and the deadline is clamped at runtime to the process's remaining share of the outer lefthook cap, which the `pre-pr-validation` job declares via `PRE_PR_OUTER_CAP_SECONDS`. The now-redundant standalone `build-all-check` lefthook job is removed so pre-push never runs two unlocked `build_all --check` processes concurrently.

### Why

Fixes #5079. `scripts/validation/pre_pr.py` never asked whether generated output was stale. CI asks it in `validate-generated-agents.yml`, so a hand-edit to a generated file cleared every local gate and failed only in CI. Measured cost on PR #5059: round-cap wiring hand-edited into `src/copilot-cli/skills/pr-autofix/SKILL.md` (generated from `.claude/commands/pr-autofix.md`) passed 26 of 26 skill tests locally, and CI then showed the generator stripping all 43 lines. The feature would have merged as a silent no-op.

## Specification References

| Type | Reference | Description |
|------|-----------|-------------|
| **Issue** | Fixes #5079 | pre_pr.py does not run the generator staleness checks locally |
| **Rule** | `.claude/rules/generated-artifacts.md` | Generator order: sync before build; negative-control requirement |
| **Rule** | `.claude/rules/ci-scripts.md` MUST 10-13, 16 | Exit on `main(argv)`, fail loud, examined counts, corpus evidence, timeout sizing |
| **ADR** | ADR-035 | Exit-code contract: 0 ok, 1 logic, 2 config, 3 external |

## Changes

- `scripts/validation/check_generated_staleness.py` (new): runs the two generator checks in contract order under one aggregate budget (`_GATE_BUDGET_SECONDS = 420.0`) shared across the rows, so the gate's worst case plus the 30s termination grace (450s) is half the 15-minute lefthook cap on `pre-pr-validation`. Because the gate's clock starts only when the sequence reaches it, the deadline is additionally clamped to the process's remaining share of the outer cap, which the lefthook job declares via `PRE_PR_OUTER_CAP_SECONDS`; a spent share reports EXTERNAL without spawning, so the outer SIGKILL can never land during a generator write. On deadline expiry the child receives SIGINT (`KeyboardInterrupt`), so the build generator's snapshot-restoring cleanup (its finally block) runs; only a child that ignores the interrupt for 30s is killed, with an explicit warning. A `_Status` IntEnum carries ADR-035 exits to the CLI (drift or child-reported failure 1, absent script or bad root 2, expired child 3); the failure message directs the reader to the child's echoed output as the discriminator and gives both remedies, because the child exit contracts are ambiguous (the build check uses 2 for both config errors and staleness; the sync check uses 1 for missing, unreadable, and drifted sources alike). The clean-path line prints elapsed seconds against the budget. The registry-contract docstring cites `scripts/validation/pre_pr_sequence.py:147` with a byte-exact signature quote and divergence section.
- `scripts/validation/pre_pr_sequence.py`: registers the gate immediately after `Orphaned Build Deferrals`, so the two different questions about the build generator sit together.
- `lefthook.yml`: removes the standalone `build-all-check` pre-push job (it raced the gate's own invocation with no lock over the same owned prefixes, and its 15m lefthook `timeout:` was a SIGKILL with the skipped-cleanup corruption path) and adds `PRE_PR_OUTER_CAP_SECONDS: "900"` to `pre-pr-validation`, pinned equal to that job's timeout by a test.
- `tests/validation/test_check_generated_staleness.py` (new, 16 tests): exit codes asserted on `main(argv)` (clean 0, drift 1, absent script 2, bad root 2, expired child 3), preserved-partial-output check, generator-order isolation with a control, echo-tail truncation, and wiring tests that drive the real `run_all_validations`.
- `tests/validation/test_check_generated_staleness_termination.py` (new, 9 tests): `test_budget_plus_grace_fits_inside_the_lefthook_cap` (parses the live `lefthook.yml`, pins budget + grace inside half the cap, and pins the declared clamp value equal to the job's timeout), `test_expiry_lets_the_child_finally_run` (marker-file proof the child's cleanup runs on SIGINT), the SIGINT-ignoring kill escalation with its partial-writes warning, the non-POSIX terminate() fallback driven with a fake, spent-share EXTERNAL refusal with an unspent-cap control, and malformed-clamp warn-and-fall-back.
- `tests/validation/staleness_gate_helpers.py` (new): shared fake-repo stubs and the clamp-env hygiene fixture for the two gate test modules, split out when round-5 coverage crossed the 500-line test file-size ceiling.
- `tests/test_lefthook_integration.py`: pins the standalone job's absence with the rationale and the `pre-pr-validation` env map including the declared cap.
- `tests/validation/test_pre_pr_sequence_registry.py` (9 tests): pins the new gate into the ordered sequence literal.
- `.agents/qa/pr-5088-generated-staleness-gate-report.md`, `.agents/sessions/2026-08-15-session-5079.json` and its episode: QA rounds 1-5 and session evidence, with the negative control re-run on the round-5 head.

## Type of Change

- [x] Bug fix (non-breaking change fixing an issue)
- [ ] New feature (non-breaking change adding functionality)
- [ ] Breaking change (fix or feature causing existing functionality to change)
- [ ] Documentation update
- [x] Infrastructure/CI change
- [ ] Refactoring (no functional changes)

## Reviewer Expectations

**Review kind / scrutiny / urgency**: validation-gate review, high scrutiny. This is a push-blocking gate, so a false positive blocks every contributor.

**Focus areas**: the aggregate-budget arithmetic against the lefthook cap and the runtime clamp against the declared outer cap (both pinned by a test that parses the live config), the graceful-termination path in `_run_check` (SIGINT, 30s grace, then kill, with the escalation branch now tested), the `_Status` narrowing at the registry boundary, and the lefthook dedup (one build-all check per pre-push, inside the sequence).

## Testing

- [x] Tests added or updated
- [x] Manual testing completed
- [ ] No testing required (documentation only)

Evidence: 34 targeted tests pass (25 gate across two modules + 9 registry) plus 847 lefthook integration tests; ruff and mypy clean; the gate against the full corpus on this branch prints `generated staleness: 0 stale in 2 generator check(s) examined (0.8s of 420s budget)`, rc=0; taste (583), ruff (27), type-ignore (44), and cli-exit-contract (27) ratchets unchanged.

Negative control (required by `.claude/rules/generated-artifacts.md` MUST 2), re-run on the round-5 head so it quotes the shipped failure message: appending a marker line to the generator source `.claude/commands/pr-autofix.md` without regenerating makes the gate name the exact file PR #5059 shipped hand-edited and exit 1; reverting returns rc=0.

### Security Review

- [x] Security patterns applied: the gate spawns two repository scripts in `--check` mode via `sys.executable` with a fixed argument list, no shell; the bounded graceful termination eliminates both the demonstrated worktree-corruption path and the unbounded-stall path, and the aggregate budget plus the runtime outer-cap clamp keep the outer lefthook SIGKILL from ever being the first deadline to fire.

### Other Agent Reviews

- [x] QA verified test coverage (`.agents/qa/pr-5088-generated-staleness-gate-report.md`, rounds 1-5)

## Review rounds 2 through 5

Round 2 (`copilot-pull-request-reviewer`, fixed in `5f8fe6262`): external kill corrupted the worktree; exit codes contradicted ADR-035; canonical-source-mirror violation. Round 3 (reliability agent + Copilot, fixed in `f233933dd` and `0c1fc3b42`): unbounded child stalled callers outside the lefthook cap, fixed with graceful SIGINT-first deadlines; the standalone `build-all-check` lefthook job raced the gate's own invocation and was removed. Round 4 (Copilot, fixed in `5b5986c2e`): two independent 600s budgets exceeded the outer 15m cap's share, replaced by the aggregate 420s budget with the cap arithmetic pinned against the live `lefthook.yml`; the failure message stopped asserting staleness for ambiguous child exits. Round 5 (Copilot + spec completeness judge, fixed in `9d2acdc87`, `ce88c21b8`, `1bf59a295`): the static half-cap split could not account for time earlier validations spent on the same outer timer, fixed with the declared-cap runtime clamp; the untested kill-escalation and non-POSIX branches gained tests; the QA report's negative control was re-run on the current head. Detail on each thread.

## Author Pre-flight

- [x] Builds locally (or N/A)
- [x] Pre-PR validation passed
- [x] Lint and formatting clean (scoped to changed files)
- [x] Code follows project style guidelines and code quality standards
- [x] Self-review completed
- [x] Comments added only where the why is non-obvious
- [x] Documentation updated (if applicable)
- [x] No new warnings introduced

## Acceptance criteria

- [x] `pre_pr.py` fails when a generated artifact is stale against its source (negative control quoted in the QA report, re-run on the round-5 head)
- [x] The gate runs sync before build and stops at the first failure (isolating marker-file test plus its control)
- [x] The gate is bounded by one aggregate budget whose worst case plus grace fits inside half the lefthook `pre-pr-validation` cap, and its deadline is clamped at runtime to the process's remaining share of the cap the job declares; expiry preserves the child's cleanup (SIGINT-first, marker-file proof; escalation and non-POSIX branches tested) and reports EXTERNAL
- [x] Exit codes follow ADR-035 (1 drift or child failure, 2 config, 3 external), asserted on `main(argv)`
- [x] The gate passes against the full corpus on this branch (rc=0, examined count and elapsed printed)
- [x] Exactly one build-all check runs per pre-push (standalone lefthook job removed; absence pinned by `tests/test_lefthook_integration.py`)
- [x] Removing the gate from `_SEQUENCE` fails a test

## Notes for Reviewers

Noticed, not fixed, kept for follow-up: `build/scripts/build_all.py:1206` prints an em dash in its `STALENESS DETECTED` message (`.claude/rules/universal.md` MUST NOT 5); the dash validator scans markdown only, so nothing catches it. On Windows, the terminate() fallback cannot preserve the child's cleanup (no SIGINT delivery; a console control event lands as SIGBREAK, which Python terminates on without running finally unless every child installs a handler); the non-POSIX branch is tested for its termination contract, and a cleanup-preserving Windows path is tracked as a follow-up.

## Related Issues

Fixes #5079
Refs #5059 (the measured incident)

### CODE REVIEW

File: scripts/validation/check_generated_staleness.py [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+            cwd=repo_root,
+            capture_output=True,
+            text=True,
+            encoding="utf-8",
+            errors="replace",
+            timeout=_CHECK_TIMEOUT_SECONDS,
+            check=False,
+        )
+    except subprocess.TimeoutExpired:
+        return -1, f"{script.name} --check exceeded {_CHECK_TIMEOUT_SECONDS}s"
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply scripts/validation/check_generated_staleness.py]: Fixed in 803a7dd95. You were right, and it reproduces.

**Verified the mechanism first.** `build/scripts/build_all.py:1123-1128` is exactly as you describe:

```python
    finally:
        # #2440: ALWAYS restore on --check, including on exception paths.
        # Otherwise a generator crash mid-build leaves partial writes
        # in the caller's worktree.
        if snapshot is not None:
            _restore_owned_prefixes(repo_root, OWNED_PREFIXES, snapshot)
```

SIGKILL cannot run that block.

**Then reproduced it.** A fixed-delay sweep does not show it: 16 kills between 0.58s and 2.0s against a deliberately stale tree left the generated trees clean every time, because the write phase is a fraction of a second and a kill outside it lands either before any write or after the restore. Polling until `git status` reported the first generated write and killing at that instant does show it:

```text
attempt=1 saw_mid_write=14 killed=True left_dirty=15
     M src/copilot-cli/skills/pr-autofix/SKILL.md
    ?? src/copilot-cli/.hook-stage-oczzzsa7/
    ?? src/copilot-cli/hooks/PostToolUse/.hook-stage-nt768qpa.tmp
    ?? src/copilot-cli/hooks/PreToolUse/.hook-stage-63cum3ub.tmp
    ?? src/copilot-cli/hooks/PreToolUse/.hook-stage-67e3p46p.tmp
  -> CORRUPTION REPRODUCED: restore never ran
```

15 dirty paths: a rewritten generated `SKILL.md` plus 14 orphaned `.hook-stage-*` staging files. Narrow, real, and invisible to casual testing, which is the worst combination for something that runs on every push.

Note the clean tree cannot show this at all, whatever the timing. The generator rewrites byte-identical content there, so a partial write leaves no diff. Only a genuinely stale tree, which is precisely the case this gate exists to catch, exposes it.

**The fix: took your first option, removed the external kill.** Measured 4.43s standalone against the 600s cap, so the cap was never a routine path, and a validation gate that can corrupt its own subject is not worth a backstop that already exists one level up in the lefthook job cap, which owns the whole process tree.

I kept a cap on the other row rather than deleting the concept. `sync_plugin_lib.py --check` is a dry run that writes nothing, so killing it leaves no partial state. The asymmetry is now a per-row field in `_CHECKS` with the reason inline, and two tests pin it in both directions so neither half can rot:

- `test_build_all_carries_no_external_kill` fails if that row ever grows a cap.
- `test_the_dry_run_row_does_carry_a_cap` is its control. Without it, deleting every cap in the table would pass the first assertion and the reason would be lost.

**Partial diagnostics preserved.** The timeout branch now keeps `TimeoutExpired.stdout` and `.stderr` and appends the kill marker, rather than discarding both. `_decode` handles the bytes-or-str quirk for the same reason `scripts/validation/subprocess_runner.py:29-37` does, and says so. I did not import `_run_subprocess` from that module: its signature requires a timeout, which is the one thing the `build_all` row must not have. Covered by `test_a_killed_child_keeps_the_output_it_already_flushed`.

**Pre-existing hazard I am not fixing here, flagging instead.** `lefthook.yml:566` runs `build-all-check` under `timeout: 15m`. That kill has the same effect on the same `finally`, and it predates this PR. Out of scope for a gate-wiring change, worth its own issue.

File: scripts/validation/check_generated_staleness.py [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+
+
+def main(argv: list[str] | None = None) -> int:
+    """CLI entry point. Returns an ADR-035 exit code."""
+    args = argv if argv is not None else sys.argv[1:]
+    repo_root = Path(args[0]).resolve() if args else Path(__file__).resolve().parents[2]
+    if not repo_root.is_dir():
+        print(f"[FAIL] Invalid repository root: {repo_root}", file=sys.stderr)
+        return 2
+    return 0 if validate_generated_staleness(repo_root) else 1
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply scripts/validation/check_generated_staleness.py]: Fixed in 803a7dd95. Correct on both counts: the module's own exit table promised 2 for an absent script, and `main` returned 1 for it.

**Took the shape you named.** A `_Status` IntEnum now carries the outcome to the CLI, and the boolean narrowing happens only at the `pre_pr` registry boundary:

```python
class _Status(IntEnum):
    OK = 0
    DRIFT = 1
    CONFIG = 2
    EXTERNAL = 3
```

The value *is* the exit code, so `main` returns `int(check_generated_staleness(repo_root))` and there is no second mapping table to drift from the first. `validate_generated_staleness` stays as the registry entry point and reduces to `check_generated_staleness(repo_root) is _Status.OK`.

Mapping now:

| Outcome | Code | ADR-035 class |
| --- | --- | --- |
| No drift | 0 | ok |
| A generator reported drift | 1 | logic |
| A checked script is absent, or the root is not a directory | 2 | config |
| A bounded child was killed on timeout | 3 | external |

The distinction is operator-facing, not cosmetic. Exit 1 tells the reader to regenerate and commit. Exit 2 tells them the gate could not run and the script needs restoring, where regenerating would be the wrong action. Exit 3 tells them the tree was never scored at all.

**Tests updated with the negative controls.**

- `test_an_absent_script_is_a_config_error_not_drift` asserts 2, replacing the assertion that pinned the defect at 1.
- `test_a_killed_child_is_an_external_error` asserts 3, driving a stub that sleeps past a 1s cap.
- `test_a_root_that_is_not_a_directory_is_a_config_error` asserts 2.
- `test_build_all_staleness_exits_one` and `test_sync_drift_exits_one` keep the drift path pinned at 1, so the widening cannot swallow the original signal.

All assert on `main(argv)` rather than a helper's return value, per `.claude/rules/ci-scripts.md` MUST 10: a helper can report failure correctly while nothing converts it into a process exit, and only the CLI-level assertion catches that.

```text
$ uv run pytest tests/validation/test_check_generated_staleness.py -q
................                                                         [100%]
16 passed
```

File: scripts/validation/check_generated_staleness.py [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+    """Return True when no generator's output is stale relative to its sources.
+
+    Entry point matching the ``validate_*(repo_root) -> bool`` contract used by
+    ``pre_pr_sequence``. Fails closed when a checked script is absent: a silent
+    skip would defeat the point of registering the gate.
+
+    Prints the examined count next to the failure count so a caller can tell
+    "no drift across two checks" from "nothing was checked"
+    (.claude/rules/ci-scripts.md MUST 11 and MUST 12).
+    """
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply scripts/validation/check_generated_staleness.py]: Fixed in 803a7dd95. The docstring said "matching the `validate_*(repo_root) -> bool` contract used by `pre_pr_sequence`" with no path and a paraphrase, which is the exact shape `.claude/rules/canonical-source-mirror.md` rejects.

It now cites the path you named and quotes the adapter signature character for character:

```python
    Canonical source of the consumed contract:
    ``scripts/validation/pre_pr_sequence.py:147``, whose adapter signature reads
    verbatim (quoted at column 0 so the 99-character original is reproduced
    byte for byte rather than wrapped to fit an indent):

def _root_only(validator: Callable[[Path], bool]) -> Callable[[Path, argparse.Namespace], bool]:

    So the registry accepts exactly ``Callable[[Path], bool]``. Different than
    canonical: this module's own CLI keeps the richer ``_Status``, because
    ADR-035 distinguishes drift (1) from a missing script (2) and a killed
    child (3) and a boolean cannot carry that. The narrowing happens here and
    nowhere else, so the registry stays uniform while the CLI stays honest.
```

Two notes on the details.

**The quote sits at column 0 on purpose.** The canonical line is 96 characters. An eight-space docstring indent pushed it to 104 and tripped ruff E501, and the two ways out of that are wrapping the line or dedenting it. Wrapping would have broken the character-for-character requirement to satisfy a line-length rule, so the byte-exact copy won and the indent lost. The parenthetical says why, so the next reader does not "fix" the indentation and silently break the quote.

**Verified programmatically rather than by eye:**

```text
canonical line 147: 'def _root_only(validator: Callable[[Path], bool]) -> Callable[[Path, argparse.Namespace], bool]:'
len: 96
quoted verbatim in module: True
```

**Added the divergence section the rule requires.** The claim is no longer bare parity: this module diverges from the consumed contract by keeping `_Status` at the CLI while handing the registry a `bool`, and the docstring names that under "Different than canonical" with the reason. Per the rule, an "aligned with X" claim with no divergence section, where the implementation does diverge, is the failure mode, so stating the parity without stating the narrowing would have left a second defect behind the first.

File: scripts/validation/check_generated_staleness.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (rjmurillo) [reply scripts/validation/check_generated_staleness.py]: Correction on the commit reference: the fix described above was authored in a session whose push never landed, so `803a7dd95` does not exist on this branch. The same fix is now actually on the branch as `5f8fe6262` (pushed head `365fe61d1`): the `build_all` row carries no external kill, the sync row keeps its 600s cap, and `test_build_all_carries_no_external_kill` plus `test_the_dry_run_row_does_carry_a_cap` pin the asymmetry in both directions. The timeout branch preserves `TimeoutExpired` partial output (`test_a_killed_child_keeps_the_output_it_already_flushed`). Resolving.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/validation/check_generated_staleness.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (rjmurillo) [reply scripts/validation/check_generated_staleness.py]: Correction on the commit reference: `803a7dd95` never reached this branch (its authoring session's push did not land). The `_Status` IntEnum fix described above is now on the branch as `5f8fe6262` (pushed head `365fe61d1`): drift 1, absent script or bad root 2 (config), killed child 3 (external), all asserted on `main(argv)` in `test_an_absent_script_is_a_config_error_not_drift`, `test_a_root_that_is_not_a_directory_is_a_config_error`, and `test_a_killed_child_is_an_external_error`. Resolving.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/validation/check_generated_staleness.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (rjmurillo) [reply scripts/validation/check_generated_staleness.py]: Correction on the commit reference: `803a7dd95` never reached this branch. The canonical-source-mirror fix described above is on the branch as `5f8fe6262` (pushed head `365fe61d1`): the docstring cites `scripts/validation/pre_pr_sequence.py:147`, quotes the 96-character `_root_only` signature byte for byte at column 0, and carries the divergence section (CLI keeps `_Status`, registry receives `bool`). Verified against the branch's own `pre_pr_sequence.py`, where line 147 holds that signature. Resolving.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/validation/pre_pr_sequence.py [hunk 4]
```diff
@@ -236,6 +237,14 @@ def _run_copilot_routing_exclusions(repo_root: Path, _args: argparse.Namespace)
...
     # Issue #2770.
     _Gate("Orphaned Build Deferrals", _run_orphaned_build_deferrals),
+    # The gate above reads deferral comments inside build_all.py's source. This
+    # one asks the separate question CI asks: is the generated tree stale
+    # against its inputs. Runs sync_plugin_lib.py --check then
+    # build/scripts/build_all.py --check, in the order
+    # .claude/rules/generated-artifacts.md requires. Both are read-only.
+    # Issue #5079: without it, a hand-edit to a generated file cleared every
+    # local gate and the generator silently reverted it in CI (PR #5059).
+    _Gate("Generated Artifact Staleness", _root_only(validate_generated_staleness)),
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply scripts/validation/pre_pr_sequence.py:247]: Fixed in `0c1fc3b42` (pushed head `8473ce3a7`) by taking the first option: the standalone `build-all-check` job is removed from `lefthook.yml`. With the gate inside `pre-pr-validation`, that job was pure duplication, and deleting it also removes its 15m lefthook `timeout:`, which was a SIGKILL with the same skipped-`finally` corruption path this PR eliminated from the gate itself. `tests/test_lefthook_integration.py` now pins the job's absence with the rationale so it cannot silently return (847 tests pass). Exactly one `build_all --check` runs per pre-push, inside the sequence, under the gate's graceful 600s deadline.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/validation/check_generated_staleness.py [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+    proc = subprocess.Popen(
+        [sys.executable, str(script), "--check"],
+        cwd=repo_root,
+        stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE,
+        text=True,
+        encoding="utf-8",
+        errors="replace",
+    )
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply scripts/validation/check_generated_staleness.py:234]: /fp This repository's floor is `requires-python = ">=3.14"` (pyproject.toml), with a 3.10 hook-portability syntax floor enforced by `scripts/validation/validate_python_syntax.py`; both are far above the 3.6 availability boundary this rule guards, so `Popen(encoding=...)` is universally available here (`.claude/rules/python.md`, Version Compatibility Reviews).

---
_Generated by [Claude Code](https://claude.ai/code)_

[the remark this answers is not in the document]
Comment (rjmurillo) [reply scripts/validation/check_generated_staleness.py:234]: /fp Same basis as the sibling finding: the repository floor is `requires-python = ">=3.14"` with a 3.10 hook-portability syntax floor, so `Popen(errors=...)` is universally available (`.claude/rules/python.md`, Version Compatibility Reviews).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/validation/check_generated_staleness.py [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+# before build". Each row is (label, repo-relative script path, timeout
+# seconds or None). Changing the order breaks the contract this module exists
+# to honor.
+#
+# Both rows carry a deadline sized for a loaded machine per ci-scripts.md
+# MUST 16, enforced by _run_check's graceful termination (SIGINT first, so
+# build_all's snapshot-restoring ``finally`` runs; see the module docstring).
+_CHECKS: tuple[tuple[str, tuple[str, ...], float], ...] = (
+    ("sync_plugin_lib.py --check", ("scripts", "sync_plugin_lib.py"), 600.0),
+    ("build_all.py --check", ("build", "scripts", "build_all.py"), 600.0),
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply scripts/validation/check_generated_staleness.py]: Fixed in `5b5986c2e` (pushed head `54fcf58d5`), taking your first option: the two per-row budgets are replaced by one aggregate `_GATE_BUDGET_SECONDS = 420.0` shared across the rows, so the gate's worst case (420s + 30s grace = 450s) is half the 15m `pre-pr-validation` cap, leaving the other half for the rest of the sequence (measured ~130s). An exhausted budget reports EXTERNAL without spawning the next child. `test_budget_plus_grace_fits_inside_the_lefthook_cap` parses the live `lefthook.yml`, extracts the job's `timeout:`, and asserts `budget + grace <= cap / 2`, so a future cap reduction or budget increase breaks loudly instead of reopening the outer-kill path. Resolving.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/validation/check_generated_staleness.py [hunk 7]
```diff
@@ -0,0 +1,307 @@
...
+                f"[FAIL] {label} was killed on timeout; the tree was never "
+                f"scored. Examined {examined} of {len(_CHECKS)} checks.{unrun}",
+                file=sys.stderr,
+            )
+            return _Status.EXTERNAL
+        print(
+            f"[FAIL] {label} reported staleness (exit {exit_code}). "
+            f"Examined {examined} of {len(_CHECKS)} checks.{unrun}",
+            file=sys.stderr,
+        )
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply scripts/validation/check_generated_staleness.py:355]: Addressed in `5b5986c2e` (pushed head `54fcf58d5`), with a deliberate boundary: the failure message no longer asserts staleness or prescribes regeneration unconditionally. It now directs the reader to the child's echoed output as the discriminator and gives both remedies ("if it reports staleness or drift, regenerate and commit; otherwise fix the error the check itself reported; regenerating is not the remedy for a configuration or source failure").

What I did not do is map child exit codes onto distinct `_Status` values, and the module docstring now records why: the child contracts are ambiguous at this boundary. `build_all.py` uses 2 for both configuration errors and staleness, and `sync_plugin_lib.py` uses 1 for missing, unreadable, and drifted sources alike, so any per-cause mapping here would assert precision the children do not provide, which is the confident-incorrectness shape `canonical-source-mirror.md` exists to prevent. The faithful fix is for the child CLIs to expose a dedicated drift exit code, at which point this gate can consume it; that is a child-contract change outside this PR's scope. Resolving on the message fix plus the recorded rationale; happy to reopen if you want the child-contract change tracked as an issue.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/validation/check_generated_staleness.py [hunk 8]
```diff
@@ -0,0 +1,346 @@
...
+    script needs restoring, where the drift remedy (regenerate and commit)
+    would be the wrong action. A silent skip would defeat the point of
+    registering the gate.
+
+    Prints the examined count next to the failure count so a caller can tell
+    "no drift across two checks" from "nothing was checked"
+    (.claude/rules/ci-scripts.md MUST 11 and MUST 12).
+    """
+    examined = 0
+    deadline = time.monotonic() + _GATE_BUDGET_SECONDS
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply scripts/validation/check_generated_staleness.py]: Fixed in `9d2acdc87` (pushed head `1bf59a295`), taking your first option: the deadline is now based on the pre-PR process's remaining outer budget. The `pre-pr-validation` job declares its own cap via `PRE_PR_OUTER_CAP_SECONDS: "900"` in `lefthook.yml`, and `_clamped_budget` computes `min(_GATE_BUDGET_SECONDS, outer_cap - elapsed_since_process_start - grace)`, where process start is captured at module import (under `pre_pr.py` that is process startup, because the sequence registry imports its validators up front). A spent share reports EXTERNAL without spawning any child, so the outer SIGKILL can never land during a generator write regardless of how much time the earlier validations consumed.

Design notes: the clamp is opt-in by environment on purpose, because a library caller such as pytest imports this module minutes before calling it, and an unconditional wall-clock clamp keyed to import time would misfire there; with the variable unset the static aggregate budget stands alone, as before. A malformed value warns loudly and falls back rather than blocking every push on a hook-config typo. The declared value is pinned equal to the job's actual `timeout:` by `test_budget_plus_grace_fits_inside_the_lefthook_cap` (which parses the live `lefthook.yml`), and the behavior is covered by `test_a_spent_outer_share_reports_external_without_running_the_child` with an unspent-cap control and the malformed-value test. For scale: the full `pre-pr-validation` job on this head measured 138.8s wall clock, so the static half-cap split also holds in practice; the clamp closes the loaded-machine tail you named. Resolving.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/validation/check_generated_staleness.py [hunk 9]
```diff
@@ -0,0 +1,346 @@
...
+    )
+    try:
+        stdout, stderr = proc.communicate(timeout=timeout)
+        return proc.returncode, (stdout or "") + (stderr or "")
+    except subprocess.TimeoutExpired as exc:
+        partial = _decode(exc.stdout) + _decode(exc.stderr)
+        if os.name == "posix":
+            proc.send_signal(signal.SIGINT)
+        else:
+            proc.terminate()
```
[the remark this answers is not in the document]
Comment (rjmurillo) [reply scripts/validation/check_generated_staleness.py:243]: Correct on the mechanism: `Popen.terminate()` is `TerminateProcess` on Windows, so the child's `finally` does not run there. Two things shipped in `9d2acdc87` (pushed head `1bf59a295`), and the full remedy is tracked as #5115 rather than folded in here.

Shipped: the non-POSIX branch is no longer untested. `test_a_non_posix_host_terminates_instead_of_signaling` drives it with a fake and pins the contract (on expiry the branch calls `terminate()` and never attempts a POSIX signal), and the escalation branch gained `test_a_child_that_ignores_the_interrupt_is_killed_with_a_warning`. The PR body's Notes for Reviewers now states the Windows limitation explicitly.

Not shipped, and why: a cleanup-preserving Windows termination is not a local change to this gate. `CTRL_BREAK_EVENT` lands in a Python child as SIGBREAK, whose default disposition terminates the process without running `finally`, so the control-event approach only preserves cleanup if every spawned checker installs a matching SIGBREAK handler; that is a cross-script contract change touching the build generator and the sync checker. Your disposable-tree alternative avoids the child contract but changes what tree the check scores. Both options are weighed in #5115 with acceptance criteria (a win32 timeout must leave the owned prefixes byte-identical, proven by a non-skipping test). On Windows today the gate's timeout behavior equals the prior standalone lefthook job's SIGKILL, so this PR is not a regression there, while POSIX (CI and the primary dev environment) gets the cleanup-preserving path. Resolving with the follow-up tracked.

---
_Generated by [Claude Code](https://claude.ai/code)_
